app/assets/javascripts/discourse/app/lib/columns.js

```diff
@@ -0,0 +1,78 @@
+/**
+ * Turns an element containing multiple children into a grid of columns.
+ * Can be used to arrange images or media in a grid.
+ *
+ * Inspired/adapted from https://github.com/mladenilic/columns.js
+ *
+ */
+export default class Columns {
+  constructor(container, options = {}) {
+    this.container = container;
+
+    this.options = {
+      columns: 3,
+      columnClass: "d-image-grid-column",
+      minCount: 2,
+      ...options,
+    };
+
+    this.excluded = ["BR", "P"];
+
+    this.items = this._prepareItems();
+
+    if (this.items.length >= this.options.minCount) {
+      this.render();
+    } else {
+      container.dataset.disabled = true;
+    }
+  }
+
+  count() {
+    // a 2x2 grid looks better for 2 or 4 items
+    if (this.items.length === 4 || this.items.length === 2) {
+      return 2;
+    }
+    return this.options.columns;
+  }
+
+  render() {
+    const count = this.count();
+    const columns = this._prepareColumns(count);
+
+    this.items.forEach((item, index) => {
+      columns[index % count].append(item);
+    });
```

Comment:
To achieve a true 'masonry' layout, I think we need to consider the aspect ratios of the images. If we just fill up the columns based on image 'count', then we can end up with one column significantly higher than the others. e.g.:

<img width="399" alt="SCR-20230605-olnn" src="https://github.com/discourse/discourse/assets/6270921/6d085a5d-44c1-483c-9928-40aa862952a7">

I think the easiest strategy will be to sum the 'height' of each column as we go, and then add the next image to the column with the smallest height. 'height' here doesn't have to be absolute - it can be relative to the width (i.e. aspect ratio). Some logic can potentially be stolen from [the topic-thumbnails theme component](https://github.com/discourse/discourse-topic-thumbnails/blob/bd495f7aca92bcb4e61a199d2a28c974110b12d1/javascripts/discourse/initializers/topic-thumbnails-init.js#L141-L168)

Reply:
Also... random thought... inserting images into the DOM 'out of order' is not ideal from a keyboard navigation or accessibility point of view. It also means that going next/previous in the lightbox goes up/down a column rather than going to the actual previous/next image.

Eventually it might be better to keep the images in the DOM as a simple list of IMGs, and position them manually using JS (but preferably keeping everything in terms of `aspect-ratio` so that it reflows nicely when the container changes size). But that sounds pretty tricky to achieve cleanly, so maybe that's a v2 goal.